# Data preparation

Helmholtz Munich Imputation Server (HMIS) accepts VCF files compressed with [bgzip](http://samtools.sourceforge.net/tabix.shtml). Please make sure the following requirements are met:

- Create a separate vcf.gz file for each chromosome.
- Variations must be sorted by genomic position.
- GRCh37 or GRCh38 coordinates are required.

!!! note
    Several \*.vcf.gz files can be uploaded at once.

## Quality Control for HRC or 1000G

HRC panel currently has a threshold for the number of strand flips allowed (see [Pipeline Overview](../pipeline/#quality-control)).

You may need to prepare your data before submitting your genotype data.

Will Rayner provides a great toolbox to prepare data: [HRC or 1000G Pre-imputation Checks](https://www.strand.org.uk/).

The main steps for HRC are:

### Download tool and sites

````sh
wget https://www.strand.org.uk/tools/HRC-1000G-check-bim-v4.2.11.zip
wget https://www.strand.org.uk/tools/HRC.r1-1.GRCh37.wgs.mac5.sites.tab.gz
````

We also provide the HRC-1000G-check-bim tool as a [Singularity/Apptainer container](https://cloud.sylabs.io/library/hmgu-itg/default/strand_tools) and [Docker image](https://hub.docker.com/repository/docker/itganalytics/strand_tools/general).

The `HRC-1000G-check-bim.pl` script is used to check [plink](https://www.cog-genomics.org/plink/) `.bim` files against a reference for strand, id names, positions, alleles, and ref/alt assignment.


### 0. Convert genotype data to plink binary format
You can use [plink](https://www.cog-genomics.org/plink/) to convert your genotype data to the plink binary file format (`bed/bim/fam`) if it is in a different format.

````sh
# Plink text format (ped/map) to plink binary format (bed/bim/fam)
plink --file <input-prefix> --make-bed --out <output-prefix>

# VCF to plink binary format
plink --vcf <input-vcf> --make-bed --out <output-prefix>
````

Skip this section if your data is already in the plink binary file format.

!!! info "Future plans"
    We are planning to provide a workflow to check your genotype data against a reference for other genotype data formats in the future.

### 1. Create a frequency file
`HRC-1000G-check-bim.pl` requires the input data frequency report to run its checks.
````sh
plink --freq --bfile <input> --out <freq-file>
````

### 2. Execute script
````sh
perl HRC-1000G-check-bim.pl -b <.bim file> -f <freq-file> -r HRC.r1-1.GRCh37.wgs.mac5.sites.tab -h
sh Run-plink.sh
````

The following reports are created by `HRC-1000G-check-bim.pl`:

- `Chromosome-<input-file>-HRC.txt`:  
- `FreqPlot-<input-file>-HRC.txt`: List of variants that exists in both input and reference file and their frequencies (for plotting purposes).  
- `Exclude-<input-file>-HRC.txt`: List of variants that are invalid and needs to be excluded.  
- `Force-Allele1-<input-file>-HRC.txt`: List of variants to flip the alleles to align with the alleles in the reference file.  
- `ID-<input-file>-HRC.txt`: List of variants that overlaps with the variants in the reference file.  
- `Position-<input-file>-HRC.txt`:  
- `Strand-Flip-<input-file>-HRC.txt`: List of variants where the stand is flipped when compared to the reference file.  
- `LOG-<input-file>-HRC.txt`: Log messages generated from running the `HRC-1000G-check-bim.pl` script.  
  
The `Run-plink.sh` is a post-processing script which is also generated by `HRC-1000G-check-bim.pl`, and running `Run-plink.sh` will generate an updated version of your input file suitable for uploading to the HMIS for imputation. 

### Create vcf using [VcfCooker](http://genome.sph.umich.edu/wiki/VcfCooker)

````sh
vcfCooker --in-bfile <bim file> --ref <reference.fasta>  --out <output-vcf> --write-vcf
bgzip <output-vcf>
````
## Additional Tools

### Convert ped/map files to VCF files

Several tools are available:
 [plink2](https://www.cog-genomics.org/plink2/),
 [BCFtools](https://samtools.github.io/bcftools) or [VcfCooker](http://genome.sph.umich.edu/wiki/VcfCooker).  

````sh
plink --ped study_chr1.ped --map study_chr1.map --recode vcf --out study_chr1
````

Create a sorted vcf.gz file using [BCFtools](https://samtools.github.io/bcftools):

````sh
bcftools sort study_chr1.vcf -Oz -o study_chr1.vcf.gz
````

### CheckVCF

Use [checkVCF](https://github.com/zhanxw/checkVCF) to ensure that the VCF files are valid. checkVCF proposes "Action Items" (e.g. upload to sftp server), which can be ignored. Only the validity should be checked with this command.

````sh
checkVCF.py -r human_g1k_v37.fasta -o out mystudy_chr1.vcf.gz
````
